<!DOCTYPE html>
<html>
	<head>
		<meta charset="utf-8">
		<title></title>
	</head>
	<body>
		<script>
			// const arr = [NaN, null, undefined, [], {}, '']
			// function test1(i){
			// 	if(Object.is(i,NaN)) {
			// 		return `${i} 是 NaN`
			// 	} else if(Object.is(i,undefined)){
			// 		return `${i} 是 undefined`
					
			// 	} else if(Object.is(i,null)){
			// 		return `${i} 是 null`
					
			// 	} else if(Object.is(i,'')){
			// 		return `'' 是 ''`
					
			// 	} else if(Object.is(JSON.stringify([]), '[]')) {
			// 		return `${JSON.stringify([])} 是 []`
					
			// 	} else if(Object.is(JSON.stringify({}), '{}')) {
			// 		return `${i} 是 {}`
					
			// 	} else {
			// 		return '啥也不是'
			// 	}
			// }
			// let aaa = test1(['s'])
			// console.log(aaa)
			// console.log(typeof undefined) //undefined
			// console.log(typeof [])  // object
			// console.log(typeof {}) //object
			// console.log(typeof '') //string
			// console.log(typeof null) //object
			// console.log(Object.is(NaN,NaN)) // true
			// console.log(Object.is(undefined,undefined)) // true
			// console.log(Object.is(null,null)) // true
			// console.log(Object.is('','')) // true
			// console.log(Object.is([],{})) // false
			// console.log(Object.is({},{})) // false
			// console.log(Array.isArray([]))
			// console.log(Array.isArray({}))
			// console.log(Object.is(JSON.stringify(['ss']), '[]'))
			// console.log(Object.is(JSON.stringify([]), '[]'))
			
			// async function a() {
			// 	return 10
			// }
			
			// const aa = a()
			// console.log(aa)
			
			// async function test1() {
			// 	console.log('test1 start') //2
			// 	await test2()
			// 	console.log('test2 end') //5
			// 	await test3()
			// 	console.log('test3 end') //7
			// }
			
			// async function test2() {
			// 	console.log('test2 start') //3
			// }
			
			// async function test3() {
			// 	console.log('test3 start') //6
			// }
			
			// console.log('script start') //1
			// test1()
			// console.log('script end') //4
			
			// async function async1 () {
			//   console.log('async1 start') // 2
			//   await async2() 
			//   console.log('async1 end') // 6
			// }
			
			// async function async2 () {
			//   console.log('async2') // 3 
			// }
			
			// console.log('script start') // 1 
			
			// setTimeout(function () { 
			//   console.log('setTimeout') // 8
			// }, 0)
			
			// async1()
			
			// new Promise (function (resolve) { 
			//   console.log('promise1') // 4
			//   resolve()
			// }).then (function () { 
			//   console.log('promise2') // 7
			// })
			
			// console.log('script end') // 5 
			
			
			class myPromise{
				state = 'pending' //状态  'pending'  'fulfilled'  'rejected'
				value = undefined  //成功后的值
				reason = undefined //失败后的原因
				
				resolveCallbacks = []  // pending状态下，存储成功的回调
				rejectCallbacks = []  // pending状态下，存储失败的回调
				
				constructor(fn) {
					const resolveHandler = (value) => {
						if (this.state === 'pending') {
							this.state = 'fulfilled'
							this.value = value
							this.resolveCallbacks.forEach(fn => fn(this.value))
						}
					}
					const rejectHandler = (reason) => {
						if (this.state === 'pending') {
							this.state = 'rejected'
							this.reason = reason
							this.rejectCallbacks.forEach(fn => fn(this.reason))
						}
					}
					
					try {
						fn(resolveHandler, rejectHandler)
					} catch (err){
						rejectHandler(err)
					}
				}
				
				then(fn1, fn2) {
					// 在pending状态下，fn1和fn2会被存储到callbacks中
					
					fn1 = typeof fn1 === 'function' ? fn1 : (v) => v
					fn2 = typeof fn2 === 'function' ? fn2 : (e) => e
					
					if (this.state === 'pending') {
						const p1 = new myPromise((resolve,reject) => {
							this.resolveCallbacks.push(() => {
								try{
									const newValue = fn1(this.value)
									resolve(newValue)
								}catch(e){
									//TODO handle the exception
									reject(e)
								}
							})
							this.rejectCallbacks.push(() => {
								try{
									const newReason = fn2(this.reason)
									reject(newReason)
								}catch(e){
									//TODO handle the exception
									reject(e)
								}
							})
						})
						return p1
					}
					if (this.state === 'fulfilled') {
						const p1 = new myPromise((resolve,reject) => {
							try{
								const newValue = fn1(this.value)
								resolve(newValue)
							}catch(e){
								//TODO handle the exception
								reject(e)
							}
						})
						return p1
					}
					if (this.state === 'rejected') {
						const p1 = new myPromise((resolve,reject) => {
							try{
								const newReason = fn2(this.reason)
								reject(newReason)
							}catch(e){
								//TODO handle the exception
								reject(e)
							}
						})
						return p1
					}
				}
				catch(fn) {
					return this.then(null, fn)
				}
			}
			myPromise.resolve = function(value) {
				return new myPromise((resolve,reject) => resolve(value))
			}
			myPromise.reject = function(reason) {
				return new myPromise((resolve,reject) => reject(reason))
			}
			myPromise.all = function(promiseList = []) {
				if(!Array.isArray(promiseList)) throw new TypeError('请传入一个数组')
				const p1 = new myPromise((resolve,reject) => {
					const result = []  //存储promiseList所有结果
					const length = promiseList.length
					let resolveCount = 0
					promiseList.forEach((p,index) => {
						p.then(data => {
							result[index] = data
							
							// resolveCount 必须在then中进行++
							resolveCount ++
							if (resolveCount === length) {
								resolve(result)
							}
						}).catch(err => {
							reject(err)
						})
					})
				})
				return p1
			}
			myPromise.race = function(promise = []) {
				if(!Array.isArray(promiseList)) throw new TypeError('请传入一个数组')
				let resolved = false //标记
				const p1 = new myPromise((resolve, reject) => {
					promiseList.forEach(p => {
						p.then(data => {
							if(!resolved) {
								resolve(data)
								resolved = true
							}
						}).catch(err => {
							if(!resolved) {
								reject(err)
								resolved = true
							}
						})
					})
				})
			}
			
			const p = new myPromise((resolve,reject) => {
				// resolve(20)
				// reject('err')
				setTimeout(function() {
					resolve(100)
				}, 1000);
			})
			const p0 = new myPromise((resolve,reject) => {
				// resolve(20)
				// reject('err')
				setTimeout(function() {
					resolve(1100)
				}, 1000);
			})
			
			const p1 = myPromise.resolve(30)
			const p2 = myPromise.resolve(40)
			const p3 = myPromise.reject('err')
			
			const p4 = myPromise.all([p,p1,p0,p2])
			console.log(p4)
			const p5 = p4.then(err => console.log(err))
			
			const p6 = myPromise.race([p, p3])
			p6.then(data => console)
			
			// const p1 = p.then(data => {
			// 	console.log('data', data)
			// 	return data + 1
			// })
			
			// const p2 = p1.then(data1 => {
			// 	console.log('data1', data1)
			// 	return data1 + 1
			// })
			
			// const p3 = p2.catch(e => console.error(e))
			
			// Promise.reject().then(() => { // 返回 rejected 状态的 promise
			//     console.log(1)
			//     throw new Error('erro1')
			// }).catch(() => { // 返回 resolved 状态的 promise
			//     console.log(2)
			// }).then(() => {
			//     console.log(3)
			// })
			
			// const a = Promise.reject()
			// a.then(() => {
			// 	console.log(1)
			// }).catch(() => { // 返回 resolved 状态的 promise
			//     console.log(2)
			// 	return 3
			// }).then((d) => {
			// 	console.log(d)
			// })
			// const a = {
			// 	length: 3,
			// 	1:'s',
			// 	2:'dd'
			// }
			// const b = Array.from(a)
			// console.log(b)
			
			// var fruits = ["Banana", "Orange", "Apple", "Mango"];
			// fruits.splice(2,0,'d');
			// console.log(fruits)
		</script>
	</body>
</html>